### PR TITLE
First draft of `apply_lora_norm_regularization`

### PR DESCRIPTION
**Pull request by jukofyork**
So now instead of `apply_max_norm_regularization` we have `apply_lora_norm_regularization`:

- The old `scale_weight_norms` parameter has been renamed to `lora_max_norm`.
- The new `lora_weight_decay` parameter applies decoupled weight-decay to the composite matrix `W = s⋅BA`.

I manually calculated the gradients needed for the loss augmented with a L2-regularisation term:

$$
L' = L + \lambda \cdot \frac{1}{2}\sum_i \lVert W_i \rVert_F^2
$$
and using:
$$
\frac{\partial(\frac{1}{2}\lVert W \rVert_F^2)}{\partial W} = W
$$
we get:
$$
\frac{\partial L'}{\partial W_i} = \lambda  \cdot  W_i
$$
and noting that:
$$
W_i = s \cdot B_i A_i
$$
we get:
$$
\frac{\partial L'}{\partial A_i} = s \cdot B_i^T\frac{\partial L'}{\partial W_i} = \lambda \cdot s \cdot B_i^TW_i
$$
$$
\frac{\partial L'}{\partial B_i} = s \cdot \frac{\partial L'}{\partial W_i}A_i^T = \lambda \cdot s \cdot W_iA_i^T
$$

But to apply "decoupled" weight-decay:

image

We use AdamW (or any other momentum-based optimiser) to optimise the original loss term:

$$
L' = L + ...
$$

and then standard SGD (*without* momentum) to optimise the regularisation term:

$$
L' = ... + \lambda \cdot \frac{1}{2}\sum_i \lVert W_i \rVert_F^2
$$

(*using the same scheduled value of `lr` for both updates!*)

In essence:

- This new "LoRA" weight-decay is a "soft" version of max-norm regularisation, and penalises `W` the larger the L2-ball.
- Max-norm regularisation is performing projected gradient descent by projecting the `W` back onto an L2-ball.
- The `lora_weight_decay` parameter is used to increase or decrease the "soft" penalty.
- The `lora_max_norm`parameter is used to set the radius of the L2-ball "hard" penally.

**NOTE**: There is some logic in the old code that came from the original [lora.py](https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py) that doesn't project small-norm `W` ***outwards*** onto the L2-ball (IIRC, the threshold is half the value of `lora_max_norm`). This is all left the same as before, as without reading the paper this came from; I have to assume this is correct...

---

### IMPORTANT:

- I haven't actually tested this code and hence why it is a "draft"  PR (I have tested the use of `model.named_parameters()` in some other regularisation code and can confirm it works though...).
- Whatever `weight_decay` you used before (ie: that was applied to `lora_A` and `lora_B` separately) should be ***squared*** for use with the new `lora_weight_decay` parameter (eg: `0.1` before would be `0.01` now).
- Be sure to set `weight_decay = 0.0` in the `[optimizer]` sub-section so as not to "double-regularise" the weights! The default value is set to `0.01` so it ***must*** be set to zero.
- The changes should have no effect on full fine-tuning, for which you should use the `weight_decay` in the `[optimizer]` section as normal (but notice now that `weight_decay` in the `[optimizer]` section is on the same scale as the new `lora_weight_decay` parameter!).
- Using a mix of full fine-tuning (via `"modules_to_save"`) and LoRA won't allow mixing the two different `weight_decay` settings currently.

---

### TODO:

- Please can somebody (@kallewoof :grin:) test if this works by setting a really high value of `lora_weight_decay`?
- Please can somebody (@tdrussell :grin:) check if my use of `model.named_parameters()` is correct and will work with models split using pipeline-parallel? The LLMs I've asked are adamant that `model.state_dict()` won't work as it makes a copy and from what I read `model.named_modules()` will likely get all the modules; rather than just those on the current pipeline stage?

Sorry I can't test it today as have some other stuff I need to work on - I'm fairly confident the maths is correct though as tested via a test driver to make sure it matches up...

### CODE REVIEW

**Comment by jukofyork** [comment]
The example config will eventually need to be changed too:

```toml
# hard clamp the magnitude of the LoRA weights
#scale_weight_norms = 1.0
```

but I wasn't sure if some of the notes from above about the scales of the new `lora_weight_decay` might also need adding to the comment, so left it for now...

**Comment by kallewoof** [comment]
> Please can somebody check if my use of model.named_parameters() is correct and will work with models split using pipeline-parallel? The LLMs I've asked are adamant that model.state_dict() won't work as it makes a copy and from what I read model.named_modules() will likely get all the modules; rather than just those on the current pipeline stage?

LLMs are always adamant about everything, that's why they are such successful liars. I would assume that the code, considering how widely used it is, works as it is and wouldn't change it without first **checking** if it is actually broken.

**Comment by jukofyork** [comment]
> > Please can somebody check if my use of model.named_parameters() is correct and will work with models split using pipeline-parallel? The LLMs I've asked are adamant that model.state_dict() won't work as it makes a copy and from what I read model.named_modules() will likely get all the modules; rather than just those on the current pipeline stage?
> 
> LLMs are always adamant about everything, that's why they are such successful liars. I would assume that the code, considering how widely used it is, works as it is and wouldn't change it without first **checking** if it is actually broken.

Yeah, sorry about not testing it - I just have a lot of other stuff to do this week and have hacked my PEFT version for the multiplication-LoRA stuff and would need to be careful to revert this (plus I have no baseline of what a "good" `weight_decay` model setting looks like and would have to run quite a few tests to just dial in a sensible value to test against).

> Edited to add: my suggestion is to see if the state_dict approach is a NOP, which I think is the concern here. We can test that by training a small model with an extreme scale_weight_norm setting and another model with no scale_weight_norm and comparing the output. When we've clarified that, assuming it is indeed a NOP, we may want to notify e.g. the kohya-ss devs that it is potentially not functional, or that it could be nice to add a comment about why it works there and that it doesn't work in situations like ours. (I have no idea what that reason would be.)

I agree it may well be that the LLMs lie to me lol! `Claude 3.5 Sonnet`, `GPT-4-Turbo` and `o1-preview` all say the same though and then offer up some code that uses `state_dict()[key].copy_()` which is a real function (I checked it wasn't a hallucinated member) to load the changes back in.

It could also just be that in place operations do work (as @tdrussell suggested), and I agree it should be tested... It was just the use of `named_parameters()` was the first thing I actually managed to get working after several frustrating hours of it not making any changes :frowning:

**Comment by kallewoof** [comment]
My GPUs are currently occupied but I will do two test runs on the current main branch next week and compare. I think setting this to a very small value should make the model learn nothing because it's regularized to smithereens.

**Comment by jukofyork** [comment]
> My GPUs are currently occupied but I will do two test runs on the current main branch next week and compare. I think setting this (`scale_weight_norms`) to a very small value should make the model learn nothing because it's regularized to smithereens.

Yeah, the old `scale_weight_norms` I've just left alone but renamed to `lora_max_norm` to better reflect it's purpose, and I'm not sure how useful this would ever be.

It's really just a the `lora_weight_decay` that needs testing against the old `weight_decay` and I *think* it should be approximately the square of `weight_decay` (so likely `lora_weight_decay = 0.01` has a similar effect to`weight_decay = 0.1`, etc).

File: train.py [hunk 1]
```diff
@@ -169,41 +169,51 @@ def evaluate(model_engine, eval_dataloaders, tb_writer, step, eval_gradient_accu
...
+
+                # Apply decoupled weight decay (indirectly) to the composite matrix BA if specified
+                # Using L = λ⋅½||W||_F²:
+                #    ∂L/∂W = λ⋅W, as ∂(½||W||_F²)/∂W = W
+                #    ∂L/∂A = s⋅Bᵗ(∂L/∂W) = λ⋅s⋅BᵗW
+                #    ∂L/∂B = s⋅(∂L/∂W)Aᵗ = λ⋅s⋅WAᵗ
+                if weight_decay > 0:                      
+                    grad_A = weight_decay * lora_scale * (B.t() @ W)
+                    grad_B = weight_decay * lora_scale * (W @ A.t())
+                    A -= lr * grad_A
```
Comment (tdrussell) [inline train.py:198]: This should work because it's an in-place operation on a nn.Parameter. I think it's normal that there is some confusion, as "A -= delta" works, but "A = A - delta" does not. Because of that, I think we should change this to be more explicit. Use "A.sub_(delta)" (same with B), and also for the code below "A.mul_(sqrt_ratio)".

Comment (jukofyork) [reply train.py]: Yeah, I agree - It's changed now.

**Comment by jukofyork** [comment]
I will look at testing this later this week and possibly add the L1 stuff:

- I can calculate the L1-norm gradients by hand and just apply soft-thresholding (see [here](https://en.m.wikipedia.org/wiki/Proximal_gradient_methods_for_learning) and [here](https://proximity-operator.net/multivariatefunctions.html)).
- The general Lp-norm gradients are way harder to calculate by hand (and IIRC actually require several steps of Newton's Method due to having no closed-form solution) and I'd have to return to using pytorch to do this (which I worry might interfere with Deepspeed...).

The only reason adding L1-regularisation might be interesting, is if you intend to merge several LoRAs and/or use [Mergekit](https://github.com/arcee-ai/mergekit/issues) with one of the "sparsity-aware" methods (eg: [TIES](https://arxiv.org/abs/2306.01708)).

**Comment by jukofyork** [comment]
One interesting thing I've found with LoRAs and weight-decay, is that you seem to need to to a much bigger value than the defaults used in pytorch (or the other optimisation wrappers):

- Using values like 0.001 or 0.01 has almost no effect.
- To actually see an effect you may need to use values > 1 (or even >> 1).

I think this may be due to the LoRA itself having a strong regularising effect (ie: [spectral regularization](https://en.m.wikipedia.org/wiki/Regularization_by_spectral_filtering)).

**Comment by jukofyork** [comment]
I'm closing this as I think it may be a fundamentally flawed idea to try to use ***any*** weight-decay for the high learning-rates / small sample sizes needed for fine-tuning (LoRA or otherwise)  with optimisers that rely on using an exponentially-weighted moving average (eg:  Adam, SGD + momentum, etc).

The reason is that the norms always end up oscillating by repeatedly over-shooting, then under-shooting:

image

This is even more pronounced for Adam because both the `beta1` and the `beta2` parameters cause oscillations at different frequencies, but even with `beta1=0` (ie: RMSprop) you still get this phenomenon...

---

I'm experimenting now with initialising `lora_A` using [nn.init.orthogonal_](https://pytorch.org/docs/stable/nn.init.html) (and `lora_B` to zeros as before), and then projecting `lora_A` back onto the unit sphere after each iteration and then applying weight-decay just to `lora_B` to shrink it back towards zero:

```python
def apply_lora_norm_regularization(model, config, current_lr):
    assert config['lora_alpha'] == config['lora_rank'], "Used `nn.init.orthogonal_` so must have: alpha = r"
    weight_decay = config['optimizer'].get('lora_weight_decay', 0.0)
    lora_B_scaler = 1.0 - (current_lr / config['optimizer']['lr']) * weight_decay if weight_decay > 0 else 1.0
    for name, param in model.named_parameters():
        if 'lora_A' in name:
            # Project each row of lora_A back onto the surface of the unit ball
            with torch.no_grad():
                param.div_(param.norm(p=2, dim=1, keepdim=True))  # TODO: Check this works for bfloat16
        elif 'lora_B' in name and lora_B_scaler < 1.0:
            # Shrink each column of lora_B back towards the origin
            with torch.no_grad():
                param.mul_(lora_B_scaler)  # TODO: Check this works for bfloat16
.
.
.
    while True:
        gc.collect()
        torch.cuda.empty_cache()
        metrics = model_engine.train_batch()
        train_dataloader.sync_epoch()
        if lora_config is not None:
            apply_lora_norm_regularization(pipeline_model, config, optimizer.param_groups[0]['lr'])
            keys_scaled, avg_norm, max_norm, norms = apply_max_norm_regularization(pipeline_model, config)
```

It still oscillates, but seems to settle down much quicker due to the constraint placed on `lora_A`.

**Comment by jukofyork** [comment]
> I'm closing this as I think it may be a fundamentally flawed idea to try to use _**any**_ weight-decay for the high learning-rates / small sample sizes needed for fine-tuning (LoRA or otherwise) with optimisers that rely on using exponentially-weighted moving average(s) (eg: Adam, SGD + momentum, etc).

This paper is relevant:

https://openreview.net/forum?id=J7V_4aauV6B

Screenshot_20241225-115743
